pymatgen/io/core.py

```diff
@@ -0,0 +1,240 @@
+"""
+This module defines the abstract interface for pymatgen InputSet and InputSetGenerator classes.
+"""
+
+import abc
+import os
+from pathlib import Path
+from collections.abc import Mapping
+from string import Template
+from typing import Union, Optional, Dict
+from zipfile import ZipFile
+from monty.json import MSONable
+from monty.io import zopen
+
+
+__author__ = "Ryan Kingsbury"
+__email__ = "RKingsbury@lbl.gov"
+__status__ = "Development"
+__date__ = "September 2021"
+
+
+class InputFile(MSONable):
+    """
+    Abstract base class to represent a single input file. Note that use
+    of this class is optional; it is possible create an InputSet that
+    does not rely on underlying Inputfile objects.
+
+    All InputFile classes must implement a __str__ method, which
+    is called by write_file.
+    """
+
+    @abc.abstractmethod
+    def __str__(self) -> str:
+        """
+        String representation of an entire input file.
+        """
+
+    def write_file(self, filename: Union[str, Path]) -> None:
+        """
+        Write the input file.
+
+        Args:
+            filename: The filename to output to, including path.
+        """
+        filename = filename if isinstance(filename, Path) else Path(filename)
+        with open(filename, "wt") as f:
+            f.write(self.__str__())
+
+    @staticmethod
+    @abc.abstractmethod
+    def from_string(contents):
+        """
+        Create an InputFile object from a string
+
+        Args:
+            contents: The contents of the file as a single string
+
+        Returns:
+            InputFile
+        """
+
+    @classmethod
+    def from_file(cls, filename: Union[str, Path]):
+        """
+        Creates an InputFile object from a file.
+
+        Args:
+            filename: Filename to read, including path.
+
+        Returns:
+            InputFile
+        """
+        filename = filename if isinstance(filename, Path) else Path(filename)
+        with zopen(filename, "rt") as f:
+            return cls.from_string(f.read())
+
+
+class InputSet(MSONable, Mapping):
+    """
+    Abstract base class for all InputSet classes. InputSet classes serve
+    as containers for all calculation input data.
+
+    All InputSet must implement a _get_inputs and from_directory method.
+    Implementing the validate method is optional.
+    """
+
+    @property
+    @abc.abstractmethod
+    def _inputs(self) -> Dict[str, Union[str, InputFile]]:
```

Comment:
I wouldn't make this a private variable. It is very strange to make a abstract property a private one.

Reply:
Agreed, and certainly they can. I feel like a private `_get_inputs` will minimize confusion for novice users though,  since the class will contain both `get_inputs` and `write_inputs` methods. At least with a private method they won't immediately see `get_inputs` when using e.g. tab completion in a notebook.

Reply:
Let's just make it public. I have regretted making things private (or fake private) more often than not. In the first place, an abstract method is a very public declaration that this is a critical signature that needs to be implemented by the developer. It doesn't make a lot of sense for such things to be private.